From d4ce89e634c5939a28b395d34bace08cac399087 Mon Sep 17 00:00:00 2001
From: Alvin Chang <alvinga@andestech.com>
Date: Wed, 6 Dec 2023 15:00:40 +0800
Subject: [PATCH 147/241] target/riscv: Refine and fix the icount trigger

1. Apply the trigger_common_match() for icount trigger.
2. Fix the logic to check the icount trigger is enabled or not.
3. Fix wrong "count--" when call itrigger_set_count(). It should be
   "--count".
4. Add trigger action for icount trigger.

Signed-off-by: Alvin Chang <alvinga@andestech.com>
---
 target/riscv/debug.c | 32 ++++++++++++--------------------
 1 file changed, 12 insertions(+), 20 deletions(-)

diff --git a/target/riscv/debug.c b/target/riscv/debug.c
index fe1d1a7542..c4bc420e6c 100644
--- a/target/riscv/debug.c
+++ b/target/riscv/debug.c
@@ -109,6 +109,8 @@ static trigger_action_t get_trigger_action(CPURISCVState *env,
         action = (tdata1 & TYPE6_ACTION) >> 12;
         break;
     case TRIGGER_TYPE_INST_CNT:
+        action = (tdata1 & ITRIGGER_ACTION);
+        break;
     case TRIGGER_TYPE_INT:
     case TRIGGER_TYPE_EXCP:
     case TRIGGER_TYPE_EXT_SRC:
@@ -812,21 +814,6 @@ itrigger_set_count(CPURISCVState *env, int index, int value)
                                    ITRIGGER_COUNT, value);
 }
 
-static bool check_itrigger_priv(CPURISCVState *env, int index)
-{
-    target_ulong tdata1 = env->tdata1[index];
-    if (env->virt_enabled) {
-        /* check VU/VS bit against current privilege level */
-        return (get_field(tdata1, ITRIGGER_VS) == env->priv) ||
-               (get_field(tdata1, ITRIGGER_VU) == env->priv);
-    } else {
-        /* check U/S/M bit against current privilege level */
-        return (get_field(tdata1, ITRIGGER_M) == env->priv) ||
-               (get_field(tdata1, ITRIGGER_S) == env->priv) ||
-               (get_field(tdata1, ITRIGGER_U) == env->priv);
-    }
-}
-
 bool riscv_itrigger_enabled(CPURISCVState *env)
 {
     int count;
@@ -834,7 +821,11 @@ bool riscv_itrigger_enabled(CPURISCVState *env)
         if (get_trigger_type(env, i) != TRIGGER_TYPE_INST_CNT) {
             continue;
         }
-        if (check_itrigger_priv(env, i)) {
+        if ((env->tdata1[i] & ITRIGGER_VS) == 0 &&
+            (env->tdata1[i] & ITRIGGER_VU) == 0 &&
+            (env->tdata1[i] & ITRIGGER_U)  == 0 &&
+            (env->tdata1[i] & ITRIGGER_S)  == 0 &&
+            (env->tdata1[i] & ITRIGGER_M)  == 0 ) {
             continue;
         }
         count = itrigger_get_count(env, i);
@@ -854,14 +845,14 @@ void helper_itrigger_match(CPURISCVState *env)
         if (get_trigger_type(env, i) != TRIGGER_TYPE_INST_CNT) {
             continue;
         }
-        if (check_itrigger_priv(env, i)) {
+        if (!trigger_common_match(env, TRIGGER_TYPE_INST_CNT, i)) {
             continue;
         }
         count = itrigger_get_count(env, i);
         if (!count) {
             continue;
         }
-        itrigger_set_count(env, i, count--);
+        itrigger_set_count(env, i, --count);
         if (!count) {
             env->itrigger_enabled = riscv_itrigger_enabled(env);
             do_trigger_action(env, i);
@@ -892,7 +883,7 @@ static void riscv_itrigger_update_count(CPURISCVState *env)
          * the count field in itrigger tdata1 register is updated.
          * And the count field in itrigger only contains remaining value.
          */
-        if (check_itrigger_priv(env, i)) {
+        if (trigger_common_match(env, TRIGGER_TYPE_INST_CNT, i)) {
             /*
              * If itrigger enabled in this privilege mode, the number of
              * executed instructions since last privilege change
@@ -986,7 +977,8 @@ static void itrigger_reg_write(CPURISCVState *env, target_ulong index,
 static int itrigger_get_adjust_count(CPURISCVState *env)
 {
     int count = itrigger_get_count(env, env->trigger_cur), executed;
-    if ((count != 0) && check_itrigger_priv(env, env->trigger_cur)) {
+    if ((count != 0) &&
+        trigger_common_match(env, TRIGGER_TYPE_INST_CNT, env->trigger_cur)) {
         executed = icount_get_raw() - env->last_icount;
         count += executed;
     }
-- 
2.34.1

